### PR TITLE
Allow specifying error type to be raised by decompose

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
As the `decompose` transform is used in various contexts, it would be nice to specify the error type that should be raised if its not possible to get a suitable decomposition for one of the operators  (i.e. devices raise `DeviceError`, transforms and gradients raise `DecompositionUndefinedError`, catalyst raises `CompileError`)

**Description of the Change:**
An error kwarg is added to `decompose` and its helper function `_operator_decomposition_gen` so that the error type can be specified. It continues to default to `DeviceError`

**Benefits:**
We don't have to duplicate many lines of code to get an identical function with a different error type

**Possible Drawbacks:**
More kwargs

**Related GitHub Issues:**
N/A

**Related Shortcut Stories:**
[[sc-61450]](https://app.shortcut.com/xanaduai/story/61450/catalyst-can-decompose-nested-operations-hybrid-op-part-2)

### CODE REVIEW

**Comment by lillian542** [comment]
[sc-61450]

File: pennylane/devices/preprocess.py [hunk 1]
```diff
@@ -253,6 +255,7 @@ def decompose(
     ] = None,
     max_expansion: Union[int, None] = None,
     name: str = "device",
+    error: Exception = DeviceError,
```
Comment (albi3ro) [inline pennylane/devices/preprocess.py:258]: Mind adding this to the list of Args too? Potentially we should break all the optional kwargs into a `Keyword Args:` section at this point.

Comment (dwierichs) [reply pennylane/devices/preprocess.py:258]: Can't comment on unchanged code, but could you also change the hard-coded ``DecompositionUndefinedError`` mentions in the `Args:` section?

**Review by dwierichs — COMMENTED** [review]
The change looks good to me, thanks @lillian542! :tada:

Would be good to update the `Args:` section as the comment details.
Also, do we want a basic test that asserts the `error` kwarg being used correctly?
And a changelog entry? :thinking:

**Comment by trbromley** [comment]
Thanks @lillian542!

> it would be nice to specify the error type that should be raised

Are you able to speak more towards that motivation? I guess this is needed for Catalyst integration, but I'm just curious to understand more.

**Comment by lillian542** [comment]
> Also, do we want a basic test that asserts the error kwarg being used correctly?
> And a changelog entry? 🤔

@dwierichs - why yes. Yes we do want those things 🙈

**Comment by lillian542** [comment]
> Are you able to speak more towards that motivation? I guess this is needed for Catalyst integration, but I'm just curious to understand more.

@trbromley - this function was originally used only in preprocessing for standard devices (which we want to raise a `DeviceError` if it gets an operation it can't use or decompose). 

As of this week, it's also used in transforms, where no device is present, and it would be better to raise the original `DecompositionUndefinedError`. For catalyst, the issue that we hit in `decompose` is whether or not it can be compiled via `qjit` for the given device, and everything in that part of the pipe line raises a `CompileError`. 

So the motivation is to keep error types consistent with what process is failing, even though at the end of the day they all come down to a `DecompositionUndefinedError` arising in different contexts.

An alternative would be to always use `DecompositionUndefinedError`. That would be accurate, but inconvenient in `try-except` blocks where we would like to, for example, try to run preprocessing and have special handling if a `DeviceError` is raised.

File: pennylane/devices/preprocess.py [hunk 2]
```diff
@@ -261,13 +264,19 @@ def decompose(
...
             should be decomposed. If an operator cannot be decomposed and is not accepted by ``stopping_condition``,
             a ``DecompositionUndefinedError`` will be raised. This replaces stopping_condition if and only if the tape has shots.
-        skip_initial_state_prep=True (bool): If ``True``, the first operator will not be decomposed if it inherits from :class:`~.StatePrepBase`.
+        skip_initial_state_prep (bool): If ``True``, the first operator will not be decomposed if it inherits
+            from :class:`~.StatePrepBase`. Defaults to ``True``.
         decomposer (Callable): an optional callable that takes an operator and implements the relevant decomposition.
             If None, defaults to using a callable returning ``op.decomposition()`` for any :class:`~.Operator` .
-        max_expansion (int): The maximum depth of the expansion.
+        max_expansion (int): The maximum depth of the expansion. Defaults to None.
+        name (str): The name of the operation or device using the transform. Used in the error message.
```
Comment (lillian542) [inline pennylane/devices/preprocess.py:277]:
```suggestion
        name (str): The name of the transform, process or device using decompose. Used in the error message. Defaults to "device".
```

**Review by albi3ro — APPROVED** [review]

**Review by dwierichs — APPROVED** [review]
Thanks @lillian542 
Just 1.5 comments regarding the docstring.

File: pennylane/devices/preprocess.py [hunk 3]
```diff
@@ -261,13 +264,19 @@ def decompose(
         stopping_condition (Callable): a function from an operator to a boolean. If ``False``, the operator
             should be decomposed. If an operator cannot be decomposed and is not accepted by ``stopping_condition``,
             a ``DecompositionUndefinedError`` will be raised.
+
+    Keyword Args:
         stopping_condition_shots (Callable): a function from an operator to a boolean. If ``False``, the operator
             should be decomposed. If an operator cannot be decomposed and is not accepted by ``stopping_condition``,
             a ``DecompositionUndefinedError`` will be raised. This replaces stopping_condition if and only if the tape has shots.
```
Comment (lillian542) [inline pennylane/devices/preprocess.py:271]:
```suggestion
            an ``Error`` will be raised. This replaces stopping_condition if and only if the tape has shots.
```

File: pennylane/devices/preprocess.py [hunk 4]
```diff
@@ -261,13 +264,19 @@ def decompose(
         stopping_condition (Callable): a function from an operator to a boolean. If ``False``, the operator
             should be decomposed. If an operator cannot be decomposed and is not accepted by ``stopping_condition``,
             a ``DecompositionUndefinedError`` will be raised.
```
Comment (lillian542) [inline pennylane/devices/preprocess.py:266]:
```suggestion
            an ``Exception`` will be raised (of a type specified by the ``error`` kwarg).
```

File: pennylane/devices/preprocess.py (hunk 3 above)
Comment (lillian542) [inline pennylane/devices/preprocess.py:271]:
```suggestion
             an ``Exception`` will be raised (of a type specified by the ``error`` kwarg). This replaces stopping_condition if and only if the tape has shots.
```

File: pennylane/devices/preprocess.py [hunk 5]
```diff
@@ -276,9 +285,9 @@ def decompose(
         The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
 
     Raises:
-        DecompositionUndefinedError: if an operator is not accepted and does not define a decomposition
-
-        DeviceError: If the decomposition enters and infinite loop and raises a ``RecursionError``.
+        Error: Type defaults to ``DeviceError`` but can be modified via keyword argument. Raised if
```
Comment (lillian542) [inline pennylane/devices/preprocess.py:288]:
```suggestion
        Exception: Type defaults to ``DeviceError`` but can be modified via keyword argument. Raised if
```

File: pennylane/devices/preprocess.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/devices/preprocess.py]: good point, updated!

File: pennylane/devices/preprocess.py [hunk 6]
```diff
@@ -253,21 +255,28 @@ def decompose(
...
             should be decomposed. If an operator cannot be decomposed and is not accepted by ``stopping_condition``,
-            a ``DecompositionUndefinedError`` will be raised.
+            an ``Exception`` will be raised (of a type specified by the ``error`` kwarg).
+
+    Keyword Args:
         stopping_condition_shots (Callable): a function from an operator to a boolean. If ``False``, the operator
             should be decomposed. If an operator cannot be decomposed and is not accepted by ``stopping_condition``,
-            a ``DecompositionUndefinedError`` will be raised. This replaces stopping_condition if and only if the tape has shots.
-        skip_initial_state_prep=True (bool): If ``True``, the first operator will not be decomposed if it inherits from :class:`~.StatePrepBase`.
+             an ``Exception`` will be raised (of a type specified by the ``error`` kwarg). This replaces stopping_condition if and only if the tape has shots.
```
Comment (lillian542) [inline pennylane/devices/preprocess.py:271]:
```suggestion
            an ``Exception`` will be raised (of a type specified by the ``error`` kwarg). This replaces stopping_condition if and only if the tape has shots.
```

**Pull request merged**